File: Frogger/obstacles.cpp
#include "obstacles.h"

Obstacles::Obstacles(Frog* frog) {

	Obstacles::frog = frog;

	level = 1;
	setStartingPositions();
	generateEnemies();
}

Obstacles::~Obstacles(){};

//Set speed of game
void Obstacles::setLevel(float level) {

	Obstacles::level = level;
}

void Obstacles::setStartingPositions(){
	
	startingPosCarX[0] = STARTING_POS_CAR_X1;
	startingPosCarX[1] = STARTING_POS_CAR_X2;
	startingPosCarX[2] = STARTING_POS_CAR_X3;

	startingPosWood[0] = STARTING_POS_WOOD_X1;
	startingPosWood[1] = STARTING_POS_WOOD_X2;
	startingPosWood[2] = STARTING_POS_WOOD_X3;

	startingPosTurtle[0] = STARTING_POS_TURTLE_X1;	
}

//Add enemies randomly
void Obstacles::generateEnemies() {
	
	int size = enemies.size();

	while (size < MAX_ENEMIES) {

		int type = rand() % 25;
		int pos = rand() % 3;
		enemies.push_back(new Wood(startingPosWood[pos], -1.0f, 0.0f));

		//Car
		if (type < 10) {
			enemies.push_back(new Car(startingPosCarX[pos], 0.0f, 0.0f));
		}
		//Turtle
		else if (type >= 10 && type <= 15) {
			enemies.push_back(new Turtle(startingPosTurtle[0], 0.0f, 100.0f));
		}
		//Wood
		else if (type > 15) {
			enemies.push_back(new Wood(startingPosWood[pos], -1.0f, 0.0f));
		}
		size = enemies.size();
	}
}

//Moves every enemy, should be called every display
void Obstacles::updatePosition() {

	int size = enemies.size();
	int move;

	for (int i = 0; i < size; i++) {
		move = rand() % 150;
		enemies[i]->collide(frog);

		if (enemies[i]->isInitialPos() && move == 1) {
			enemies[i]->move(0.0f, 0.0f, 0.0f, level); //No parameters given because they control themselves	
		}
		else if (!enemies[i]->isInitialPos()) {
			enemies[i]->move(0.0f, 0.0f, 0.0f, level);
		}
	} 
	frog->collide(frog); 
}

//Render enemies
void Obstacles::render(VSShaderLib shader) {

	int size = enemies.size();
	
	for (int i = 0; i < size; i++) {
		if (!enemies[i]->isInitialPos())
			enemies[i]->render(shader);
	}
}

void Obstacles::destroyObstacles() {
	
	float actual;
	int size = enemies.size();

	for (int i = 0; i < size; i++) {
		actual = enemies[i]->getPosition()[2];
		if (actual > SCALE_FACTOR_Z || actual < 0.0f) {
			enemies[i]->resetPosition();
		}
	}
}

File: FroggerJS/obstacles.js
function Obstacles() {
	
	this.level = 1;

	this.startingPosCarX = [];
	this.startingPosWoodX = [];
	this.startingPosTurtleX = [];

	this.enemies = [];

	this.setStartingPositions();
	this.generateEnemies();
}

Obstacles.prototype.setLevel = function(level) {

	this.level = level;
}

Obstacles.prototype.setStartingPositions = function () {

	this.startingPosCarX[0] = 6.0;
	this.startingPosCarX[1] = 9.0;
	this.startingPosCarX[2] = 12.0;

	this.startingPosWoodX[0] = 20.0;
	this.startingPosWoodX[1] = 26.0;
	this.startingPosWoodX[2] = 26.0;

	this.startingPosTurtleX[0] = 23.0;
}

Obstacles.prototype.generateEnemies = function() {

	var size = this.enemies.length;

	while (size < 10) {

		var type = Math.floor(Math.random() * 25);
		var pos = Math.floor(Math.random() * 3);

		//Car
		if (type < 7) {
			this.enemies.push(new Car(this.startingPosCarX[pos], 0.0, 0.0));
		} 
		else if (type >= 7 && type <= 15) {
			this.enemies.push(new Turtle(this.startingPosTurtleX[0], 0.0, 100.0));
		} 
		else if (type > 15) {
			this.enemies.push(new Wood(this.startingPosWoodX[pos], 0.0, 0.0));
		}
		size = this.enemies.length;
	}
}

Obstacles.prototype.updatePosition = function() {

	var size = this.enemies.length;
	var move;

	
	for(var i = 0; i < size; i++) {
		move = Math.floor(Math.random() * 150);
		this.enemies[i].collide(frog);

		if (this.enemies[i].isInitialPos() && move == 1) {
			this.enemies[i].move(this.level);
		}
		else if (!this.enemies[i].isInitialPos()) {
			this.enemies[i].move(this.level);
		}
	}
	frog.collide();
}

Obstacles.prototype.render = function() {

	var size = this.enemies.length;
	
	for (var i = 0; i < size; i++) {
		if (!this.enemies[i].isInitialPos()) {
			this.enemies[i].render();
		}	
	}
}

Obstacles.prototype.destroyObstacles = function() {

	var size = this.enemies.length;

	for (var i = 0; i < size; i++) {
		var actual = this.enemies[i].z;
		if (actual > 100.0 || actual < 0.0) {
			this.enemies[i].resetPosition();
		}
	}
}